crates/router/src/db/payment_method.rs

```diff
@@ -61,6 +72,297 @@ pub trait PaymentMethodInterface {
+        &self,
+        customer_id: &str,
+        merchant_id: &str,
+        status: common_enums::PaymentMethodStatus,
+    ) -> CustomResult<i64, errors::StorageError> {
+        let conn = connection::pg_connection_read(self).await?;
+        storage::PaymentMethod::get_count_by_customer_id_merchant_id_status(
+            &conn,
+            customer_id,
+            merchant_id,
+            status,
+        )
+        .await
+        .map_err(|error| report!(errors::StorageError::from(error)))
+    }
+
+    #[instrument(skip_all)]
+    async fn insert_payment_method(
+        &self,
+        payment_method_new: storage::PaymentMethodNew,
+        storage_scheme: MerchantStorageScheme,
+    ) -> CustomResult<storage::PaymentMethod, errors::StorageError> {
+        let conn = connection::pg_connection_write(self).await?;
+        
+        match storage_scheme {
+            MerchantStorageScheme::PostgresOnly => {
+                payment_method_new
+                .insert(&conn)
+                .await
+                .map_err(|error| report!(errors::StorageError::from(error)))
+            },
+            MerchantStorageScheme::RedisKv => {
+                let merchant_id = payment_method_new.merchant_id.clone();
+                let customer_id = payment_method_new.customer_id.clone();
+
+                let key = format!("mid_{}_cust_{}", merchant_id, customer_id);
+                let field = format!("payment_method_id_{}", payment_method_new.payment_method_id);
+
+
+                let reverse_lookup_entry = |v: String| diesel_models::ReverseLookupNew {
+                    sk_id: field.clone(),
+                    pk_id: key.clone(),
+                    lookup_id: v,
+                    source: "payment_method".to_string(),
+                    updated_by: storage_scheme.to_string(),
+                };
+
+                let lookup_id1 = format!("payment_method_{}", &payment_method_new.payment_method_id);
+                let mut reverse_lookups = vec![lookup_id1];
+                if let Some(locker_id) = &payment_method_new.locker_id{
+                    reverse_lookups.push(format!("payment_method_locker_{}", locker_id))
+                }
+
+                let results = reverse_lookups
+                    .into_iter()
+                    .map(|v| 
+                        self.insert_reverse_lookup(reverse_lookup_entry(v), storage_scheme)
+                    );
+                
+                futures::future::try_join_all(results).await?;
+
+                let storage_payment_method = diesel_models::PaymentMethod::from(&payment_method_new);
+
+                let redis_entry = kv::TypedSql {
+                    op: kv::DBOperation::Insert {
+                        insertable : kv::Insertable::PaymentMethod(payment_method_new),
+                    },
+                };
+
+                match kv_wrapper::<diesel_models::PaymentMethod, _, _>(
```

Comment:
Partiotionkey in kv_wrapper function is hardcoded to `MerchantIdPaymentIdCombination`. You can probably add MerchantIdCustomerId combination and avoid writing mid and cid prefixes manually.

Reply:
how could we avoid constructing manually, the caller has to provide all fields anyway right?